cpp/src/arrow/compute/row/grouper.h

```diff
@@ -30,6 +30,49 @@
 namespace arrow {
 namespace compute {
 
+/// \brief A segment of contiguous rows for grouping
+struct ARROW_EXPORT GroupingSegment {
+  int64_t offset;
+  int64_t length;
+  bool is_open;
+  bool extends;
+};
+
+inline bool operator==(const GroupingSegment& segment1, const GroupingSegment& segment2) {
+  return segment1.offset == segment2.offset && segment1.length == segment2.length &&
+         segment1.is_open == segment2.is_open && segment1.extends == segment2.extends;
+}
+inline bool operator!=(const GroupingSegment& segment1, const GroupingSegment& segment2) {
+  return !(segment1 == segment2);
+}
+
+/// \brief Computes grouping segments for a batch. Each segment covers rows with identical
+/// values in the batch. The values in the batch are often selected as keys from a larger
+/// batch.
+class ARROW_EXPORT GroupingSegmenter {
+ public:
+  virtual ~GroupingSegmenter() = default;
+
+  /// \brief Construct a GroupingSegmenter which receives the specified key types
+  static Result<std::unique_ptr<GroupingSegmenter>> Make(
+      const std::vector<TypeHolder>& key_types, bool nullable_keys = false,
+      ExecContext* ctx = default_exec_context());
+
+  /// \brief Return the key types of this segmenter
+  virtual const std::vector<TypeHolder>& key_types() const = 0;
+
+  /// \brief Reset this grouping segmenter
+  virtual Status Reset() = 0;
+
+  /// \brief Get the next segment for the given batch starting from the given offset
+  virtual Result<GroupingSegment> GetNextSegment(const ExecSpan& batch,
+                                                 int64_t offset) = 0;
+
+  /// \brief Get the next segment for the given batch starting from the given offset
+  virtual Result<GroupingSegment> GetNextSegment(const ExecBatch& batch,
```

Comment:
What is the difference between ExecSpan and ExecBatch and why do we need to handle both?

Reply:
An `ExecSpan` is [a view](https://github.com/apache/arrow/blob/e8e50a3ca96ec8eb2f2beb383cf9005ce1c93486/cpp/src/arrow/compute/exec.h#L348-L350) into an `ExecBatch`. The latter owns the memory of its members while the latter doesn't. Another difference is that an `ExecSpan` cannot be chunked, as its values (i.e., columns) can only be a view into an array, but `ExecBatch` can have chunked array values.

Reply:
Thanks for explaining. My next question would be how do we pick which one to use here?

Reply:
One needs `ExecBatch` when it contains chunked arrays and `ExecSpan` cannot capture the view. Since we removed support for chunked arrays, I think the `ExecBatch` API can be removed too.

Reply:
Done.